sotodlib/io/load_book.py

```diff
@@ -145,170 +96,603 @@ def load_obs_book(db, obs_id, dets=None, prefix=None, samples=None,
-                if len(bias_dest):
-                    bias_names = _check_bias_names(frame)[:_TES_BIAS_COUNT]
-                    delta = _extract_2d(frame['tes_biases'], start,
-                                        bias_dest[detset_index], dest_offset,
-                                        bias_names)
-
-                # Extract the main signal
-                if not no_signal:
-                    delta = _extract_2d(
-                        frame['signal'],
-                        start,
-                        aman['signal'],
-                        dest_offset,
-                        aman.dets.vals,
-                    )
-                
-            # How many samples were added to dest?
-            if delta > start:
-                dest_offset += (delta - start)
-                start = 0
-            else:
-                start -= delta
+            more_data &= primary_acc.append(frame['primary'], frame_offset)
 
-            if dest_offset >= samples[1] - samples[0]:
-                break
+            bias_names = _check_bias_names(frame)[:_TES_BIAS_COUNT]
+            more_data &= bias_acc.append(frame['tes_biases'], frame_offset)
+
+            # Even if no_signal, we need the det list.
+            if this_stream_dets is None:
+                this_stream_dets = _compact_list(frame['signal'].names)
+
+            # Extract the main signal
+            if not no_signal:
+                more_data &= signal_acc.append(frame['signal'], frame_offset)
+
+        if not more_data:
+            break
 
-        # Wrap up for detset ...
-        all_bias_names.extend(bias_names)
+    if times_acc is not None:
+        times_acc = times_acc.finalize() / spt3g_core.G3Units.sec
+
+    req_dets_in_stream = None
+    det_idx_in_stream = None
+    if this_stream_dets:
+        # Of the requested detectors, what ones were actually found in
+        # this stream?  Keep ordering same as the request, which
+        # should be the same as the populated data.
+        if dets is None:
+            req_dets_in_stream = this_stream_dets
+        else:
+            req_dets_in_stream = _compact_list(
+                [d for d in dets if d in this_stream_dets])
+
+        # For each loaded detector, what was its index within the stream's
+        # dets?
+        det_idx_in_stream = _compact_list([this_stream_dets.index(d)
+                                           for d in req_dets_in_stream])
+
+    stat = smurf_proc.get_status()
+    ch_info = None
+    iir_params = None
+    if stat.num_chans is None:
+        # No wiring frames probably means it's an A_* (ancil) file.
+        stat = None
+    else:
+        # This stuff will need to be regrouped, for all detectors.  So
+        # just extract individual items, restricting to dets actually
+        # used here.
+        ch_info_full = load_smurf.get_channel_info(stat)
+        ch_info = {}
+        for k, v in ch_info_full._fields.items():
```

Comment:
`get_channel_info` accepts a mask you can use to already down-select the channel information.